feat(model-ad): create typescript charts library and angular wrapper library, add boxplot chart

hallieswan opened the pull request:
## Description

Creates a new typescript charts library and an angular wrapper library. Adds a new boxplot chart using Apache ECharts.

## Related Issues

- [AG-1073](https://sagebionetworks.jira.com/browse/AG-1073)
- [AG-1085](https://sagebionetworks.jira.com/browse/AG-1085)
- [AG-1439](https://sagebionetworks.jira.com/browse/AG-1439)

## Changelog

- Add shared typescript charts library -- `shared-typescript-charts`
  - Initially generated with `nx g @nx/js:library shared-typescript-charts --bundler=tsc --unitTestRunner=jest --testEnvironment=jsdom --directory=libs/shared/typescript/charts --projectNameAndRootFormat=as-provided`, but required changes to tsconfig.base.json, .eslintrc.json, and project.json to align with monorepo setup
  - Add new boxplot chart that uses Apache ECharts to:
    - Display 1+ points per category
    - Offset points to prevent overlap of categories 
    - Change color and shape of points by category
    - Optionally display tooltips for each point and on each x-axis label
    - Display precomputed (static) boxplot summaries or calculate dynamic boxplot summaries based on provided points
- Add shared angular library that wraps the charts library -- `shared-typescript-charts-ngx`
  - Created by copying `shared-typescript-util` library, then updating the library name
  - Add boxplot directive that wraps typescript boxplot chart for use in Angular apps
    ```zsh
    nx g @nx/angular:directive boxplot --directory=/libs/shared/typescript/charts-ngx/src/lib/boxplot
    ``` 
  - Add [Storybook](https://storybook.js.org/docs/get-started/angular) to showcase charts

    ```zsh
    # add storybook to sage-monorepo
    nx add @nx/storybook
   
    # add storybook to shared-typescript-charts-ngx project
    nx g @nx/angular:storybook-configuration shared-typescript-charts-ngx --interactionTests=false --generateStories=true --configureStaticServe=true
    ```
  - Add [testing-library](https://testing-library.com/docs/angular-testing-library/intro/) and [jest-dom](https://github.com/testing-library/jest-dom) to simplify testing
     
     ```zsh
     yarn add --dev @testing-library/angular
     yarn add --dev @testing-library/jest-dom
     ```

## Preview

### Agora style boxplot

Static summaries, 1 point per boxplot: 

https://github.com/Sage-Bionetworks/sage-monorepo/assets/26949006/99007b90-a468-44fc-b7da-30cf5b227a72

### Model-AD style boxplot

1+ point per boxplot separated by category:

https://github.com/Sage-Bionetworks/sage-monorepo/assets/26949006/83f7d69c-063a-469a-9886-7a30c5394127

Dynamically calculated summaries: 

https://github.com/Sage-Bionetworks/sage-monorepo/assets/26949006/fa2604d5-045b-4292-99d0-4f1ddf2237f3

[AG-1073]: https://sagebionetworks.jira.com/browse/AG-1073?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AG-1085]: https://sagebionetworks.jira.com/browse/AG-1085?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AG-1439]: https://sagebionetworks.jira.com/browse/AG-1439?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

hallieswan commented on libs/shared/typescript/charts/.eslintrc.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "extends": ["../../../../.eslintrc.json"],
+  "ignorePatterns": ["!**/*"],
+  "env": {
+    "browser": true
+  },
```
Specify that the charts library is intended to run in browser environments, so browser-specific interfaces are defined, e.g. `HTMLDivElement`

tschaffter replied:
@hallieswan Do you know if the Apache ECharts library supports SSR? We could bump the priority of enabling SSR up before investing time using Apache ECharts.

hallieswan replied:
The [Apache ECharts documentation](https://apache.github.io/echarts-handbook/en/how-to/cross-platform/server/) indicates that the library supports SSR, but I'm not sure how much effort it would be to implement.

tschaffter replied:
We can test this support soon when we setup the project for SSR.

hallieswan commented on libs/shared/typescript/charts-ngx/src/lib/boxplot/boxplot.directive.spec.ts:
```diff
@@ -0,0 +1,73 @@
+};
+
+describe('BoxplotDirective', () => {
+  it('should render no data placeholder when no data is passed', async () => {
+    await renderTestComponent({ points: [] });
+    expect(
+      screen.getByLabelText('No data is currently available.')
+    ).toBeVisible();
+  });
```
The chart is rendered as a canvas element, so I used the `aria-label` generated by [Apache ECharts](https://echarts.apache.org/handbook/en/best-practices/aria/) to confirm that the correct chart has been displayed. 

I initially wanted to add [Playwright component testing](https://playwright.dev/docs/test-components) to use screenshots for visual comparison / testing, but component testing is not available for Angular yet and is still an experimental feature for other frameworks (e.g. React).

hallieswan commented on libs/shared/typescript/charts-ngx/src/lib/boxplot/boxplot.directive.stories.ts:
```diff
@@ -0,0 +1,62 @@
+const meta: Meta<BoxplotDirective> = {
+  component: BoxplotDirective,
+  title: 'directives/sageBoxplot',
+  argTypes: {
+    pointTooltipFormatter: {
+      control: { type: 'function' },
+    },
+  },
```
Must specify that `pointTooltipFormatter` is a function for Storybook to correctly pass the argument along to the Angular component

hallieswan commented on libs/shared/typescript/charts-ngx/src/lib/boxplot/boxplot.directive.stories.ts:
```diff
@@ -0,0 +1,62 @@
+export default meta;
+type Story = StoryObj<BoxplotDirective>;
+
+export const NoData: Story = {
+  args: {
+    points: [],
+  },
+};
```
<img width="1521" alt="stories_noData" src="https://github.com/Sage-Bionetworks/sage-monorepo/assets/26949006/cf609d2e-41b7-4649-8847-88eb7de545bc">

hallieswan commented on libs/shared/typescript/charts-ngx/src/lib/boxplot/boxplot.directive.stories.ts:
```diff
@@ -0,0 +1,62 @@
+  },
+};
+
+export const StaticSummary: Story = {
+  args: {
+    points: staticBoxplotPoints,
+    summaries: staticBoxplotSummaries,
+    xAxisCategoryToTooltipText: {
+      CAT1: 'Category 1',
+      CAT2: 'Category 2',
+      CAT3: 'Category 3',
+      CAT4: 'Category 4',
+    },
+    xAxisTitle: 'BRAIN REGION',
+    yAxisTitle: 'LOG 2 FOLD CHANGE',
+    yAxisMax: 100,
+    yAxisMin: -100,
+    title: 'AD Diagnosis (males and females)',
+    pointTooltipFormatter: (pt: CategoryPoint) => {
+      return `Value: ${pt.value}`;
+    },
+  },
+};
```
<img width="1523" alt="stories_staticBoxplot" src="https://github.com/Sage-Bionetworks/sage-monorepo/assets/26949006/042ceae0-5a9a-41cf-82e0-63b0d0000e6c">

hallieswan commented on libs/shared/typescript/charts-ngx/src/lib/boxplot/boxplot.directive.stories.ts:
```diff
@@ -0,0 +1,62 @@
+  },
+};
+
+export const DynamicSummary: Story = {
+  args: {
+    points: dynamicBoxplotPoints,
+    xAxisTitle: 'MODEL',
+    yAxisTitle: '#objects/sqmm',
+    pointTooltipFormatter: (pt: CategoryPoint) =>
+      `${pt.pointCategory}: ${pt.value}`,
+  },
+};
```
<img width="1521" alt="stories_dynamicBoxplot" src="https://github.com/Sage-Bionetworks/sage-monorepo/assets/26949006/8ef47a47-db03-4a82-bc5a-b1b4a5370d97">

hallieswan commented on libs/shared/typescript/charts/src/lib/boxplot-chart/boxplot-chart.ts:
```diff
@@ -0,0 +1,299 @@
+    const dataForStaticBoxplotSummaries = summaries
+      ? addXAxisValueToBoxplotSummaries(summaries, xAxisCategories)
+      : undefined;
+    const dataForDynamicBoxplotTransforms =
+      formatCategoryPointsForBoxplotTransform(
+        dataForScatterPoints,
+        xAxisCategories
+      );
```
ECharts can calculate the boxplot summaries based on the provided points data, but the data needs to be formatted as a 2d array where each row contains all the values for a particular category, rather than an array of objects.

hallieswan commented on libs/shared/typescript/charts/src/lib/models/boxplot.ts:
```diff
@@ -0,0 +1,55 @@
+
+export type CategoryAsValuePoint = CategoryPoint & {
+  /* x-axis category for this point as a number based on
+  1-based indexing of the sorted, unique x-axis category values
```
```suggestion
  1-based indexing of the unique x-axis category values
```

hallieswan commented on libs/shared/typescript/charts/src/lib/models/boxplot.ts:
```diff
@@ -0,0 +1,55 @@
+
+export type CategoryAsValueBoxplotSummary = CategoryBoxplotSummary & {
+  /* x-axis category for this point as a number based on
+  1-based indexing of the sorted, unique x-axis category values */
```
```suggestion
  1-based indexing of the unique x-axis category values */
```

hallieswan commented on libs/shared/typescript/charts/src/lib/utils/boxplot-utils.spec.ts:
```diff
@@ -0,0 +1,364 @@
+        ['a', 'b', 'c'],
+        []
+      );
+      valuePoints.forEach((valuePt, index) => {
+        expect(valuePt.xAxisValue).toBeCloseTo(updatedPoints[index].xAxisValue);
+      });
```
There is a slight rounding error, so use `toBeCloseTo` to compare the xAxisValues rather than directly comparing the objects with `toEqual`

hallieswan commented on libs/shared/typescript/charts/src/lib/utils/chart-utils.ts:
```diff
@@ -0,0 +1,60 @@
+    },
+  };
+
+  // notMerge must be set to true to override any existing options set on the chart
+  chart.setOption(noDataOptions, true);
```
https://github.com/Sage-Bionetworks/sage-monorepo/assets/26949006/7a2522fd-1881-4f58-a255-8ae2e972138b

hallieswan commented on libs/shared/typescript/charts/src/lib/x-axis-label-tooltips/x-axis-label-tooltips.ts:
```diff
@@ -0,0 +1,66 @@
+import { ECharts } from 'echarts';
+
+/**
+ * Manually add tooltip to chart x-axis labels.
+ * See https://github.com/apache/echarts/issues/19616 and https://github.com/apache/echarts/issues/11903/
+ * Based on https://codepen.io/plainheart/pen/jOGBrmJ.
+ */
+export class XAxisLabelTooltips {
```
ECharts doesn't currently support tooltips on x-axis labels, so manually add them to the plot.

hallieswan commented on package.json:
```diff
@@ -44,6 +44,8 @@
     "ngx-countup": "13.1.0",
     "ngx-typed-js": "2.1.1",
     "probot": "12.2.8",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
```
These are Storybook dependencies

hallieswan commented on tsconfig.base.json:
```diff
@@ -10,89 +10,49 @@
-        "libs/shared/typescript/util/src/index.ts"
-      ],
+      "@sagebionetworks/openchallenges/util": ["libs/openchallenges/util/src/index.ts"],
+      "@sagebionetworks/shared/charts": ["libs/shared/typescript/charts/src/index.ts"],
+      "@sagebionetworks/shared/charts-ngx": ["libs/shared/typescript/charts-ngx/src/index.ts"],
```
I added paths for the two new libraries. The other changes are just related to auto-formatting.

hallieswan commented:
Rebased to resolve merge conflict and added jittering of points within categories: 

https://github.com/Sage-Bionetworks/sage-monorepo/assets/26949006/7ab36cce-91b2-4935-b6f4-4bda64da91c6

tschaffter commented on libs/shared/typescript/charts-ngx/src/lib/boxplot/boxplot.directive.stories.ts:
```diff
@@ -0,0 +1,62 @@
+export default meta;
+type Story = StoryObj<BoxplotDirective>;
+
+export const NoData: Story = {
```
```suggestion
export const noData: Story = {
```
Variables are usually lowerCamelCase in TypeScript, or sometimes UPPER_SNAKE_CASE (constants). The same comment applies to other variables created in this PR. Is this a Storybook convention by chance?

hallieswan replied:
Yes, the [Storybook docs](https://storybook.js.org/docs/writing-stories#defining-stories) recommend using UpperCamelCase for story exports.

tschaffter reviewed (approved):
LGTM. I made a few comments.

hallieswan commented on libs/shared/typescript/charts-ngx/.eslintrc.json:
Thanks for sharing the docs! For consistency with other sage-monorepo projects, I'll change this library to use the `-angular` suffix so it's easier to identify the libraries to update if we switch to using the `ngx-` prefix in the future.

sagely1 commented on libs/shared/typescript/charts-ngx/src/lib/boxplot/boxplot.directive.ts:
```diff
@@ -0,0 +1,65 @@
+} from '@sagebionetworks/shared/charts';
+
+@Directive({
+  selector: '[sageBoxplot]',
```
might be a nitpick, but should this be named sageBoxPlot?

hallieswan replied:
The [Wikipedia article](https://en.wikipedia.org/wiki/Box_plot) references both "box plot" and "boxplot", but "box plot" seems to be more common, so I'll change the name to `sageBoxPlot`!

hallieswan replied:
Good point @tschaffter! I'll stick with "boxplot".

sagely1 commented on libs/shared/typescript/charts-ngx/src/lib/boxplot/boxplot.directive.ts:
```diff
@@ -0,0 +1,65 @@
+export class BoxplotDirective implements OnChanges, OnInit, OnDestroy {
+  boxplot: BoxplotChart | undefined;
+
+  @Input({ required: true }) points: CategoryPoint[] = [];
+  @Input() summaries: CategoryBoxplotSummary[] | undefined;
+  @Input() title = '';
+  @Input() xAxisTitle = '';
+  @Input() yAxisTitle = '';
+  @Input() yAxisMin: number | undefined;
+  @Input() yAxisMax: number | undefined;
+  @Input() xAxisCategoryToTooltipText: Record<string, string> | undefined;
+  @Input() pointTooltipFormatter: undefined | ((pt: CategoryPoint) => string);
```
Inputs look good! I would expect that in the future, to make it more reusable we may want to override some styles (colors, font sizes and the like).  Looks good for now, though!

sagely1 reviewed (approved):
Looks good!

hallieswan commented on libs/shared/typescript/charts-ngx/README.md:
```diff
@@ -0,0 +1,7 @@
+# shared-charts-ngx
```
Changed to `shared-typescript-charts-angular`
